src/net_processing.cpp

```diff
@@ -2194,6 +2215,203 @@ void PeerManagerImpl::SendBlockTransactions(CNode& pfrom, Peer& peer, const CBlo
     m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::BLOCKTXN, resp));
 }
 
+/**
+ * Special handling for unconnecting headers that might be part of a block
+ * announcement.
+ *
+ * We'll send a getheaders message in response to try to connect the chain.
+ *
+ * The peer can send up to MAX_UNCONNECTING_HEADERS in a row that
+ * don't connect before given DoS points.
+ *
+ * Once a headers message is received that is valid and does connect,
+ * nUnconnectingHeaders gets reset back to 0.
+ */
+void PeerManagerImpl::HandleFewUnconnectingHeaders(CNode& pfrom, Peer& peer,
+        const std::vector<CBlockHeader>& headers)
+{
+    const CNetMsgMaker msgMaker(pfrom.GetCommonVersion());
+
+    LOCK(cs_main);
+    CNodeState *nodestate = State(pfrom.GetId());
+
+    nodestate->nUnconnectingHeaders++;
+    // Try to fill in the missing headers.
+    if (MaybeSendGetHeaders(pfrom, m_chainman.ActiveChain().GetLocator(m_chainman.m_best_header), peer)) {
+        LogPrint(BCLog::NET, "received header %s: missing prev block %s, sending getheaders (%d) to end (peer=%d, nUnconnectingHeaders=%d)\n",
+            headers[0].GetHash().ToString(),
+            headers[0].hashPrevBlock.ToString(),
+            m_chainman.m_best_header->nHeight,
+            pfrom.GetId(), nodestate->nUnconnectingHeaders);
+    }
+    // Set hashLastUnknownBlock for this peer, so that if we
+    // eventually get the headers - even from a different peer -
+    // we can use this peer to download.
+    UpdateBlockAvailability(pfrom.GetId(), headers.back().GetHash());
+
+    // The peer may just be broken, so periodically assign DoS points if this
+    // condition persists.
+    if (nodestate->nUnconnectingHeaders % MAX_UNCONNECTING_HEADERS == 0) {
+        Misbehaving(peer, 20, strprintf("%d non-connecting headers", nodestate->nUnconnectingHeaders));
+    }
+    return;
+}
+
+bool PeerManagerImpl::CheckHeadersAreContinuous(const std::vector<CBlockHeader>& headers) const
+{
+    uint256 hashLastBlock;
+    for (const CBlockHeader& header : headers) {
+        if (!hashLastBlock.IsNull() && header.hashPrevBlock != hashLastBlock) {
+            return false;
+        }
+        hashLastBlock = header.GetHash();
+    }
+    return true;
+}
+
+bool PeerManagerImpl::MaybeSendGetHeaders(CNode& pfrom, const CBlockLocator& locator, Peer& peer)
+{
+    const CNetMsgMaker msgMaker(pfrom.GetCommonVersion());
+
+    const auto current_time = GetTime<std::chrono::seconds>();
```

Comment:
Thanks, I tried to make this switch in the last commit but it took me several tries to get it to compile, so please review and let me know if I'm using `NodeClock` correctly!

Reply:
Dropping the std::atomic seems to have made things work with both clang and g++, so I think this is probably right now (thanks to @dongcarl and @sipa for helping me sort through this!).

Reply:
Interesting, the `std::atomic` version did work for me locally (compiled with clang). What you have now looks good to me!

Reply:
@MarcoFalke I think I'll probably give up on trying to use `std::atomic` because all my efforts seem, but if you feel like writing the code and can share a diff that compiles I'd be happy to include it.

Reply:
I presume the issue was g++-8 in https://cirrus-ci.com/task/5966736871653376?logs=ci#L2017

I tried my proposal (calling `NodeSeconds{}`) and it worked on top of your commit.

```diff
# git log -1 && git diff 
commit fa24d37213cede7136ed34d481091b84642c62a6 (HEAD)
Author: Suhas Daftuar <sdaftuar@gmail.com>
Date:   Tue Jun 28 10:53:02 2022 -0400

    Replace GetTime() with NodeClock in MaybeSendGetHeaders()
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index a75e3bd7f..6fc3009cb 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -358,7 +358,7 @@ struct Peer {
     std::deque<CInv> m_getdata_requests GUARDED_BY(m_getdata_requests_mutex);
 
     /** Time of the last getheaders message to this peer */
-    std::atomic<NodeClock::time_point> m_last_getheaders_timestamp{};
+    std::atomic<NodeClock::time_point> m_last_getheaders_timestamp{NodeSeconds{}};
 
     Peer(NodeId id)
         : m_id{id}
```

An equivalent would be to call `NodeClock::time_point{}` instead.

Reply:
I think the problem here is `std::atomic<T>` requires `T` to be trivially copyable, but unlike `duration`, `time_point` isn't implemented as trivially copyable (and none of `time_point`, `duration`, or the underlying `rep` are specified as being trivially copyable either). Since we already assume `duration` is trivially copyable, last time I messed with this I wrote a wrapper, something like:

```c++
template<typename TP>
class atomic_time_point {
private:
    using dur = typename TP::duration;
    std::atomic<dur> d;
public:
    atomic_time_point() noexcept : d{dur{0}} { }
    atomic_time_point(const atomic_time_point&) = delete;
    atomic_time_point(const TP& t) noexcept : d{t.time_since_epoch()} { }
    TP load() { return TP{d.load()}; }
    TP operator=(const TP& t) { d = t.time_since_epoch(); return t; }
};
using AtomicNodeTime = atomic_time_point<NodeClock::time_point>;
using AtomicNodeSeconds = atomic_time_point<NodeSeconds>;
```

which would let you write `AtomicNodeTime m_last_getheaders_timestamp{};` without relying on `time_point` implementation details.

Reply:
> time_point isn't implemented as trivially copyable

Can you add references to your claim that they are not trivially copyable?

See also the cpp program which passes on gcc7/clang5:

```cpp
#include <chrono>

template <class T>
void CanUseInAtomic() {
    static_assert(std::is_trivially_copyable_v<T>);
    static_assert(std::is_copy_constructible_v<T>);
    static_assert(std::is_move_constructible_v<T>);
    static_assert(std::is_copy_assignable_v<T>);
    static_assert(std::is_move_assignable_v<T>);
}

int main() {
    CanUseInAtomic<bool>();
    CanUseInAtomic<std::chrono::seconds>();      // duration
    CanUseInAtomic<std::chrono::system_clock::time_point>();  // time_point
}
```



https://godbolt.org/z/3nKc43cPv

Reply:
Hmm, fair point. `static_assert(std::is_nothrow_default_constructible_v<T>);` catches it for clang 5, but not for gcc 7.1 for what that's worth.

But I don't think the pre-C++20 behaviour of default `std::atomic` initialization is ever what we want anyway? Per https://en.cppreference.com/w/cpp/atomic/atomic/atomic

> 1) The default constructor is trivial: no initialization takes place other than [zero initialization](https://en.cppreference.com/w/cpp/language/zero_initialization) of static and thread-local objects. [std::atomic_init](https://en.cppreference.com/w/cpp/atomic/atomic_init) may be used to complete initialization.

that will just leave things undefined, like `int x;` ?

(Anyway, no need for further changes in this PR, as far as I can see)

Reply:
I can't say I'm exactly following everything you guys are talking about, but @MarcoFalke thank you -- I updated the PR and the compilers all seem happy now. :). Will squash that last fixup commit now...